src/utils/common.ts

```diff
@@ -290,33 +292,57 @@ export async function estimateTotalGasRequiredByUnsignedTransaction(
 }
 
 /**
- * Create an unsigned transaction of a fillRelay contract call
- * @param spokePool The specific spokepool that will populate this tx
- * @param destinationTokenAddress A valid ERC20 token (system-wide default is UDSC)
- * @param simulatedRelayerAddress The relayer address that relays this transaction
- * @returns A populated (but unsigned) transaction that can be signed/sent or used for estimating gas costs
+ * Create an unsigned transaction to fill a relay. This function is used to simulate the gas cost of filling a relay.
+ * @param spokePool A valid SpokePool contract instance
+ * @param fillToSimulate The fill that this function will use to populate the unsigned transaction
+ * @returns An unsigned transaction that can be used to simulate the gas cost of filling a relay
  */
-export async function createUnsignedFillRelayTransaction(
+export async function createUnsignedFillRelayTransactionFromFill(
   spokePool: SpokePool,
-  destinationTokenAddress: string,
-  simulatedRelayerAddress: string
+  fillToSimulate: Fill
 ): Promise<PopulatedTransaction> {
-  // Populate and return an unsigned tx as per the given spoke pool
-  // NOTE: 0xBb23Cd0210F878Ea4CcA50e9dC307fb0Ed65Cf6B is a dummy address
-  return await spokePool.populateTransaction.fillRelay(
-    "0xBb23Cd0210F878Ea4CcA50e9dC307fb0Ed65Cf6B",
-    "0xBb23Cd0210F878Ea4CcA50e9dC307fb0Ed65Cf6B",
-    destinationTokenAddress,
-    "10",
-    "10",
-    "1",
-    "1",
-    "1",
-    "1",
-    "1",
-    [],
+  const {
+    message,
+    recipient: recipientAddress,
+    relayer: relayerAddress,
+    amount: amountToRelay,
+    depositor: depositorAddress,
+  } = fillToSimulate;
+  const provider = spokePool.provider;
+
+  const isMessageEmpty = message.length === 0 || message === "0x";
+  // Only continue sanity checks if the message is not empty
+  if (!isMessageEmpty) {
+    const isRecipientAnAddress = ethers.utils.isAddress(recipientAddress);
+    const isRelayerAnAddress = ethers.utils.isAddress(relayerAddress);
+    if (!isRecipientAnAddress || !isRelayerAnAddress) {
+      throw new Error("Could not simulate message fill. Recipient address or relayer address is not a valid address");
+    }
+    const [isRecipientAContract, relayerBalanceOfToken] = await Promise.all([
+      isContractAddress(recipientAddress, provider),
+      provider.getBalance(fillToSimulate.destinationToken),
```

Comment:
this is incorrect, this is the relayer's `ETH` balance. The token balance is `ERC20.balanceOf`

Reply:
This is a holdover from a less recent change. It's resolved here: https://github.com/across-protocol/sdk-v2/pull/427/commits/860c2c3c1f3235064e37af1fcb25d5b3d728d077

Reply:
ah https://github.com/across-protocol/sdk-v2/commit/860c2c3c1f3235064e37af1fcb25d5b3d728d077#r129265169

Reply:
Done and Done https://github.com/across-protocol/sdk-v2/pull/427/commits/72e4c2b53b8c8bf6bfeb752ffa7f2965247798d0 🙂